### ISSUE TITLE
LoraTap SC400 curtain switch (TUYA Zigbee model 'TS130F')

### ISSUE THREAD

**Issue by joancf** [Koenkk/zigbee2mqtt#4257]
# New device

I have a LoraTap SC400 curtain switch (TUYA Zigbee model 'TS130F')
https://es.aliexpress.com/item/4001190275372.html?spm=a2g0s.9042311.0.0.56d863c0iyL2vS
## What happened
From scratch it says that it is not recognized (i'm on branch master up to date)
After setting the debug mode
I saw the device was 
'TS130F'
and checked some of the messages
##  I declared a new device that seems to work
```
 {
        zigbeeModel: ['TS130F'],
        model: 'TS130F',
        vendor: 'TuYa',
        description: 'Curtain/blind switch',
        whiteLabel: [
             {vendor: 'loraTap', model: 'SC400'},
        ],
        supports: 'open, close, stop, position',
        fromZigbee: [fz.tuya_curtain, fz.ZMCSW032D_cover_position_tilt],
        // toZigbee: [tz.tuya_curtain_control], //, tz.tuya_curtain_options, tz.cover_position_tilt],
 	toZigbee: [tz.cover_state, tz.cover_position_tilt_inverted],
        meta: {configureKey: 1, multiEndpoint: true},
        configure: async (device, coordinatorEndpoint) => {
            const endpoint = device.getEndpoint(1);
            await bind(endpoint, coordinatorEndpoint, ['closuresWindowCovering']);
	 },
    },
```
using the commands
`zigbee2mqtt/DEVICE/set`
`{"position":"XXX"} `  
with XXX from 0 to 100
and 
`{"state":"CLOSE"}  {"state":"OPEN"}  {"state":"STOP"}`

But I don't understand the internals of the to/from zigbee nor the async funcion..  So even that it is working I should say that I'm not sure of the solution I did so I will not post a pull request ;-) 

I still have some messages in the log like:

```Received Zigbee message from 'z_per', type 'attributeReport', cluster 'genBasic', data '{"65506":27,"appVersion":68}' from endpoint 1 with groupID 0
No converter available for 'TS130F' with cluster 'genBasic' and type 'attributeReport' and data '{"65506":27,"appVersion":68}'
```

The messages that work are:
``` 
Received Zigbee message from 'z_per', type 'attributeReport', cluster 'closuresWindowCovering', data '{"61440":1,"currentPositionLiftPercentage":0}' from endpoint 1 with groupID 0
```

What is missing: 
Calibration.
I can't calibrate the device and the opening/close times are too short.

Best and thanks for the package

**Comment by joancf** [comment]
I don't know how to calibrate the device
I do not have any idea of the messages sent. Anybody nows how to calibrate these devices, even if it's not the same model I can try to send messages to check if any has an effect!
Thanks

**Comment by joancf** [comment]
I found this from tuya...
![image](https://user-images.githubusercontent.com/2848623/92302837-abbe3880-ef6f-11ea-9f59-9591b7cd885a.png)

but i don't know what to do with it...;-)

**Comment by cobirnm** [comment]
I have a similar curtain switch a have the exact same problem. I have tried contacting the seller but up until now the only response is that it will contact the factory.  In the tuya app it seems to exist the calibration function. I don't have a tuya hub, so I also need to figure out what is the command to start calibration. Have you also tried asking your seller?

**Comment by janmo0** [comment]
> I found this from tuya...
> ...
> 
> but i don't know what to do with it...;-)

Where you found this table?

**Comment by joancf** [comment]
i went into IOT.tuya.com and i registered as developer.
Then just moving there around there is some information.
I think that you can try to develop your own module and also debug the protocol but I was not able to reach that point.
I think that the best solution is to have a tuya zigbee router and perform the calibration from the app while sniffing the traffic, but I don't have the router :-(

**Comment by joancf** [comment]
As @janmo0  indicates, I think is better to  continue this discussion in:


so I close this issue, even that is s not solved 
Duplicate of [https://github.com/dresden-elektronik/deconz-rest-plugin/issues/3134](url)

**Comment by joancf** [comment]
I think that the discussion here: 
Loratap Tuya Smart Life ZigBee 3.0 Curtain Blind Switch for Roller Shutter Electric motor dresden-elektronik/deconz-rest-plugin#3134
has more information

**Comment by djfanatix** [comment]
I have the same device, and indeed cannot find a way to get into calibration mode.
I got following data from my Homey

--

2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ZigBeeDevice has been inited
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ------------------------------------------
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] Node: 9af200f1-9106-4bf7-9a79-91a39f188d71
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] - Battery: false
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] - Endpoints: 0
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] -- Clusters:
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] --- zapp
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] --- genBasic
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- 65503 : 


2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- 65506 : 27
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- cid : genBasic
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- sid : attrs
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- appVersion : 68
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] --- genGroups
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- 65533 : 1
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- cid : genGroups
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- sid : attrs
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- nameSupport : 0
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] --- genScenes
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- 65533 : 1
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- cid : genScenes
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- sid : attrs
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- count : 0
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- currentScene : 0
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- currentGroup : 0
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- sceneValid : 0
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- nameSupport : 0
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] --- genOnOff
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- 20480 : 0
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- 32769 : 0
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- cid : genOnOff
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- sid : attrs
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- onTime : 90
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- offWaitTime : 90
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- onOff : null
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] --- genTime
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- cid : genTime
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- sid : attrs
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- localTime : 0
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] --- genOta
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- cid : genOta
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- sid : attrs
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] --- closuresWindowCovering
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- 10 : null
2020-09-11 14:53:24 [log] [ManagerDrivers] [curtain] [0] ---- 19 : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- 20 : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- 4338 : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- 32768 : 0
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- 61440 : 1
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- 61441 : 1
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- 61442 : 0
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- 61443 : 100
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- cid : closuresWindowCovering
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- sid : attrs
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- currentPositionLiftPercentage : 100
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- physicalCloseLimitTiltDdegree : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- physicalCloseLimitLiftCm : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- currentPositionLiftCm : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- windowCoveringType : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- currentPositionTiltPercentage : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- installedOpenLimitLiftCm : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- installedClosedLimitLiftCm : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- installedOpenLimitTiltDdegree : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- configStatus : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ---- intermediateSetpointsLift : null
2020-09-11 14:53:25 [log] [ManagerDrivers] [curtain] [0] ------------------------------------------

**Comment by joancf** [comment]
I think that the only way is to have a tuya zigbee gateway and use the zigbbee dongle as a sniffer. Then calibrate the device using the app while sniffing the messages to discover which are the codes sent.

**Comment by djfanatix** [comment]
I ordered a Zigbee Gateway on Amazon. (I could always return it ;-) ), so will see

**Comment by joancf** [comment]
I asked Tuya technical support they answer, 
'' Hello, it is not possible to use this way. The device must be connected to Tuya Cloud through a gateway to work normally. We cannot provide more technical support through zigbeee2mqtt. Thank you for your understanding.''

**Comment by slashbv** [comment]
I have ordered the same for testing and it arrived today. I have found this model in devices.js but somehow it is not working. Maybe it's something different.
// Lonsonho
    {
        fingerprint: [
            {modelID: 'TS130F', manufacturerName: '_TZ3000_egq7y6pr'},
        ],
        model: '11830304',
        vendor: 'Lonsonho',
        description: 'Curtain switch',
        supports: 'open, close, stop, position',
        fromZigbee: [fz.tuya_curtain, fz.ignore_basic_report],
        toZigbee: [tz.tuya_curtain_control, tz.tuya_curtain_options],
    }

**Comment by slashbv** [comment]
were you able to make it work ?

**Comment by joancf** [comment]
I don't have a gateway, let's see if @djfanatix  is able to do it!

**Comment by djfanatix** [comment]
Hi,I have sent the gateway back and also the loratap stuff.
I have now shelly's in place, which are much more reliable.

I guess the calibration is in the Tuya specific zigbee header.

**Comment by slashbv** [comment]
I have shellys too but I want to move to zigbee becasue I have changed 2 routers so far. After a while I have to reboot them regularly ... maybe becasue of too many devices.

**Comment by slashbv** [comment]
I have also one Zemismart ZM-CSW032-D, it works well now but I wanted to see if this loratap is better. At least it doesn't have an annoying light shining all the time as zemismart :)

**Comment by janmo0** [comment]
> I have also one Zemismart ZM-CSW032-D, it works well now but I wanted to see if this loratap is better. At least it doesn't have an annoying light shining all the time as zemismart :)

We're you able to calibrate the Zenismart devices?

**Comment by slashbv** [comment]
yes, check this: https://github.com/Koenkk/zigbee2mqtt/issues/3216

**Comment by barsurfer** [comment]
> yes, check this: #3216

Hi, so you basically changed the time in configuration.yaml with
time_close: <sec>
time_open: <sec>
properties?

**Comment by barsurfer** [comment]
> 2x time

Hi, can you please share how you've changed  `2x time`?

**Comment by slashbv** [comment]
I didn't. I only use time_close and time_open.

**Comment by barsurfer** [comment]
@slashbv thanks for your reply! I have SC400 and trying to configure it at the moment. I've tried the same approach as for CSW032, using fz.ZMCSW032D_cover_position_tilt and setting time_close and time_open in configuration.yaml but it seems to not work since I've it is reporting position -1 every time I'm trying to perform actions on it. Will try adding more logs to understand the possible issue. 
Is it reporting the current position properly in your case when using time_close / time_open for ZMCSW032D?

**Comment by slashbv** [comment]
this time_close and time_open was implemented because the switch is not reporting position well, after adding this I have no problems so far, just I was not able to open to a specific position in automation but I tried only one as I didn't have time to investigate more, I might have done something wrong.

**Comment by djfanatix** [comment]
I also haven’t configured it. I used a Homey btw

**Comment by barsurfer** [comment]
@slashbv  Got it! Thanks! 

For Loratap I was able to successfully pair it with HA making the changes to `devices.js` with the following config:

```
    {
        zigbeeModel: ['TS130F'],
        fingerprint: [
            {modelID: 'TS130F', manufacturerName: '_TZ3000_8kzqqzu4'}
        ],,
        model: 'SC400',
        vendor: 'Loratap',
        description: 'Curtain switch',
        supports: 'open, close, stop, position',
        fromZigbee: [fz.cover_position_tilt, fz.ignore_basic_report],
        toZigbee: [tz.cover_state, tz.cover_position_tilt],
        meta: {configureKey: 1, coverInverted: true},
    }
```

and also making a mapping like` 'SC400': [cfg.cover_position, cfg.sensor_cover],`. 
After these changes, I could see it and control it in the devices, like: 
![image](https://user-images.githubusercontent.com/11078520/94550554-94b3f300-025c-11eb-9c8c-f97e0c3f4bab.png)

The issue happened when I installed the second switch I have for the doors, which are 1.5x-2x higher than the window. 
It seems like the default time for the switch is around 10 secs, so for the window, it works great as it takes around 9 seconds to be closed, but in case of the door, it doesn't work well (since it will take around 18 seconds to fully close it). Without calibration or having the ability to set times I can't use it in a normal way :( Using `fz.ZMCSW032D_cover_position_tilt` with `time_close / time_open` properties doesn't work well too since it reports -1 for the position in my case. 

And of course, I'm not excited about buying a Tuya hub for just a calibration purpose...
Would be great if anyone can assist in solving this issue.

**Comment by slashbv** [comment]
I also hope someone will help with this model

**Comment by ronsrons** [comment]
I was also trying desperately to calibrate it.
Perhaps this could take us one step closer, there's a bit more info on the deconz page: 
https://github.com/dresden-elektronik/deconz-rest-plugin/issues/3134
but they also didn't succeed.

**Comment by barsurfer** [comment]
I'm gonna try ordering a tuya hub and then sniffing the traffic but it could take some time for me to wait for a hub to be delivered.

**Comment by 1subu** [comment]
I also have a bunch of theese switches and the same problems. I got a tuya hub and sniffered the zigbee traffic.
Here are the results:

Cluster | CMD | Attributes | Data Type | Value | Description

(0x0102) | Write Attributes (0x02) | 0xf002 | 8-Bit (0x30) | 1 (0x01) | Motor reversal on
(0x0102) | Write Attributes (0x02) | 0xf002 | 8-Bit (0x30) | 0 (0x00) | Motor reversal off
(0x0102) | Write Attributes (0x02) | 0xf001 | 8-Bit (0x30) | 0 (0x00) | Start Calibration
(0x0102) | Write Attributes (0x02) | 0xf001 | 8-Bit (0x30) | 1 (0x01) | End Calibration
(0x0006) | Write Attributes (0x02) | 0x8001 | 8-Bit (0x30) | 0 (0x00) | Light Mode  1
(0x0006) | Write Attributes (0x02) | 0x8001 | 8-Bit (0x30) | 1 (0x01) | Light Mode  2
(0x0006) | Write Attributes (0x02) | 0x8001 | 8-Bit (0x30) | 2 (0x02) | Light Mode  3

So is someone able to add it to the zigbee2mqtt converters?

**Comment by 1subu** [comment]
I got it working, quick an dirty. I tested only the calibration.
If someone wants to test it, here are my changes:

In cluster.ts add attributes to cluster closuresWindowCovering
			tuyaMovingState: { ID: 0xf000, type: dataType_1.default.enum8 },
			tuyaCalibration: { ID: 0xf001, type: dataType_1.default.enum8 },
			tuyaMotorReversal: { ID: 0xf002, type: dataType_1.default.enum8 },
In cluster.ts add attributes to cluster genOnOff
                        tuyaBacklightMode: { ID: 0x8001, type: dataType_1.default.enum8 },

In fromZigbee.js add to converter cover_position_tilt:
			if (msg.data.hasOwnProperty('tuyaMovingState')) {	
				const value = msg.data['tuyaMovingState'];	
				const movingLookup = {0: 'up', 1: 'stop', 2: 'down'};	
				result.moving = movingLookup[value];	
			}
			if (msg.data.hasOwnProperty('tuyaCalibration')) {
				const value = msg.data['tuyaCalibration'];	
				result.moving = value;	
			}
			if (msg.data.hasOwnProperty('tuyaMotorReversal')) {
				const value = msg.data['tuyaMotorReversal'];	
				result.moving = value;	
			}

In toZigbee.js add converters
    tuya_cover_calibration: {
        key: ['tuya_cover_calibration'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('closuresWindowCovering', {tuyaCalibration: value});
            return {state: {tuya_cover_calibration: value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('closuresWindowCovering', ['tuyaCalibration']);
        },
    },
    tuya_cover_reversal: {
        key: ['tuya_cover_reversal'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('closuresWindowCovering', {tuyaMotorReversal: value});
            return {state: {tuya_cover_reversal: value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('closuresWindowCovering', ['tuyaMotorReversal']);
        },
    },
    tuya_backlight_mode: {
        key: ['tuya_backlight_mode'],
        convertSet: async (entity, key, value, meta) => {
            await entity.write('genOnOff', {tuyaBacklightMode: value});
            return {state: {tuya_backlight_mode: value}};
        },
        convertGet: async (entity, key, meta) => {
            await entity.read('genOnOff', ['tuyaBacklightMode']);
        },
    },

in devices.js
	// LoraTap SC400 curtain switch
    {
        zigbeeModel: ['TS130F'],
		fingerprint: [{modelID: 'TS130F', manufacturerName: '_TZ3000_8kzqqzu4'}],
        model: 'TS0601_curtain_switch',
        vendor: 'TuYa',
        description: 'Curtain/Blind switch',
        supports: 'open, close, stop, position',
        fromZigbee: [fz.cover_position_tilt],
        toZigbee: [tz.cover_state, tz.cover_position_tilt, tz.tuya_cover_calibration, tz.tuya_cover_reversal, tz.tuya_backlight_mode],
        whiteLabel: [
            {vendor: 'LoraTap', model: 'SC400'},
        ],
    },

i hope i forgot nothing

**Comment by demonio2332** [comment]
can you share the files please, thanks

**Comment by barsurfer** [comment]
@1subu I will try it in the nearest future. Can you describe the process of calibration?

**Comment by slashbv** [comment]
nice work, I will test in the next days. How can we do the calibration ?

**Comment by 1subu** [comment]
Calibration:
1. Fully open the cover
2. Start calibration
3. Push the down-button at the switch
4. When the cover is fully closed, push the stop-button
5. End calibration.

That´s all. The switch will store the time

**Comment by slashbv** [comment]
what do we have to to to start and stop calibration process? What about going up ? Usually the time to go up is longer than going down.

**Comment by joancf** [comment]
I think is thought for curtains that move horizontally...
The first problem is that you need to install it with the blinds fully open, otherwise, you must press the down button and wait the seconds you think it will take to do the full movement even if it's already closed. 
And I think that it is better to give always some extra seconds, to avoid problems if you don't close/open it completely for some times, because the different speeds for going up and down create accumulative errors that are only reset when you open/close it completely.

**Comment by 1subu** [comment]
> 
> 
> what do we have to to to start and stop calibration process? What about going up ? Usually the time to go up is longer than going down.

I made a fork of zigbee-herdsman and the zigbee-herdsman-converters. 
https://github.com/1subu/zigbee-herdsman/tree/tuya-ts130f-cluster-attributes
https://github.com/1subu/zigbee-herdsman-converters/tree/add-converters-for-tuya-TS130F-RollerShutter-switch
Clone it and you can test it i. e with MQTT.fx

Publish 'on' or 'off' to 
zigbee2mqtt/my_curtain/set/tuya_cover_calibration

Report any errors

**Comment by slashbv** [comment]
I managed to add it but I didn't manage to start calibration. Also in HA the controls are switched meaning when I press close button it does open and open does close. Same for position 0 should be closed and 100 opened.
There is this message when Z2M starts: warn  2020-10-20 22:44:19: Supported device 'TS0601_curtain_switch' has no Home Assistant mapping
and this error:
error 2020-10-20 22:44:53: Failed to setup reporting for '0xbc33acfffe47afe2' - Error: ConfigureReporting 0xbc33acfffe47afe2/1 genOnOff([{"attribute":"onOff","minimumReportInterval":0,"maximumReportInterval":300,"reportableChange":0}], {"timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":true,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null}) failed (Timeout - 23665 - 1 - 3 - 6 - 7 after 10000ms)
    at Timeout._onTimeout (/app/node_modules/zigbee-herdsman/dist/utils/waitress.js:46:35)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)

**Comment by barsurfer** [comment]
@slashbv haven't tried this yet but looks like there should be an additional change to add a mapping for HA here:
https://github.com/Koenkk/zigbee2mqtt/blob/master/lib/extension/homeassistant.js#L2089
like I did initially, can you try something like that:
`'TS0601_curtain_switch': [cfg.cover_position, cfg.sensor_cover]` 
Probably to switch the controls you could try `tuya_cover_reversal`?

**Comment by demonio2332** [comment]
Zigbee2MQTT:info  2020-10-21 16:24:20: MQTT publish: topic 'zigbee2mqtt/bridge/log', payload '{"message":"Publish 'set' 'calibration' to 'persiana_despacho' failed: 'Error: Unknown attribute 'tuyaCalibration', specify either an existing attribute or a number'","meta":{"friendly_name":"persiana_despacho"},"type":"zigbee_publish_error"}'

**Comment by 1subu** [comment]
The new cluster attributes are merged and are available in dev branch.
I updated the PR for the new converters and the device
I can't help with HA. I am using openHAB. But i added exposes. Maybe that will help

**Comment by slashbv** [comment]
I have tested again with the latest dev. In Home Assistant the controls works well (down is down, up is up) but the position 100% should be open and 0% should be closed. Could you change this? The calibration works as you described its stores the closing time but not the opening so it assumes the closing time is the as as opening. Backlight works fine. Tilt does nothing so maybe you can take it out as it ads extra controls in Home Assistant. 
![image](https://user-images.githubusercontent.com/59421121/97214792-e13f0f80-17c3-11eb-938c-2028020ee8ba.png)

**Comment by 1subu** [comment]
Zigbee officially expects 'open' to be 0 and 'closed' to be 100. So the implementation ist right. I know HA is a bit special. OpenHAB supports the zigbee standard.
Did you try motor_reversal option? Maybe this works for HA
An other option may be, to add Device specific configuration in devices.yaml. Try to add "invert_cover: true" to your device.
The used converter "cover_position_tilt" should support this. 
For cover position i used the existing converter and standard is lift and tilt. But tilt values are only add to mqtt message when the device supports it. 
Is it not possible to customize the controls in HA?

**Comment by slashbv** [comment]
invert_cover: true works fine but considering Koenkk's the comment here it seems it is the other way around: open is 100 and close is 0 https://github.com/Koenkk/zigbee2mqtt/issues/4412#issuecomment-703168083
Regarding tilt this device does not support it and there is no point in having tilt implemented.

**Comment by slashbv** [comment]
if you do exposes: [e.cover_position(), ...  instead of exposes: [e.cover_position_tilt(), ... in devices.js seems to be fine in Home Assistant
![image](https://user-images.githubusercontent.com/59421121/97499560-5c442980-196e-11eb-9238-7f075e41ba37.png)

**Comment by vide** [comment]
@1subu THANK YOU VERY MUCH, just received a TS130F, was going to patch by myself and I saw your commit, came here...and found the solution to the blind not fully opening! Now after calibrating it with what you said works beautifully. Thanks again!

**Comment by 1subu** [comment]
@slashbv I will change this with the next PR, thanks

**Comment by vide** [comment]
How are you supposed to control the backlight? I tried with

```
mosquitto_pub -t zigbee2mqtt/0x842e14fffe9a5673/set/backlight_mode -m off
``` 

and also `low` as the payload but the blue light is still on. No errors in Z2M logs thou. Anyway this device has enough specific things that it is worth having a wiki page for it. I could prepare a PR for that if you don't mind @1subu but I still need to figure out more info on how to operate it (backlights specifically)

**Comment by 1subu** [comment]
Payload for backlight is LOW, MEDIUM or HIGH.
There are 3 light modes in the Tuya app.
Unfortunately my switch-model has no backlight, so i could not test the function. But the cluster attributes are set correctly.
It would be great if you make a wiki page for the device.
Please report if backlight works and a picture wit the backlight would be nice

**Comment by vide** [comment]
Ok, just tried. With `MEDIUM` and `HIGH` I get the stop/pause button turned on, like this:

![image](https://user-images.githubusercontent.com/270267/97584084-eedcdb00-19f7-11eb-9c83-1e56594edb49.png)


while with `LOW` the stop/pause button turns off. The blue light that indicates the device is paired with a zigbee controller, stays always on no matter what.

**Comment by slashbv** [comment]
@vide where did you buy this switch from ? Mine doesn't have backlight either but a light is shown when I press a button and I can control the intensity of that light with LOW, MEDIUM or HIGH.

**Comment by vide** [comment]
@slashbv it's [this one](https://www.aliexpress.com/item/4000014271536.html?spm=a2g0s.9042311.0.0.66ca63c0HYbfxO) I think they changed the actual model over time thou (because in the reviews someone said it worked with zigbee2mqtt out of the box and it wasn't the case with my device)

**Comment by slashbv** [comment]
@vide yours is a different device different manufacturer check this: https://github.com/Koenkk/zigbee2mqtt/issues/3216

**Comment by vide** [comment]
@slashbv actually it's what I thought I bought but it identifies itself as TS130F, that's why I'm here. Unfortunately I don't have the box anymore to check the device name there but it identifies itself as a TS103F.

**Comment by slashbv** [comment]
@vide everything is working except the lighting ?

**Comment by vide** [comment]
@slashbv yep, everything is working..well I guess somehow the backlight is working in a manner, there is just no difference between HIGH and MEDIUM and LOW is actually an off, and it applies to the "stop" light and not the always-on blue light but calibration works, position works and obviously open/close work.

**Comment by slashbv** [comment]
interesting, I have the old version and there is no calibration and also no always in blue light, only always on white light for the stop. The blue stays on only when it's not paired.

**Comment by barsurfer** [comment]
@slashbv I have the same as OP (from the same seller), had a chance today to test the dev branch, so there are my observations: 
1) calibration works;
2) controls for tilt -  agree they should be removed since not supported;
3) strange issue with the UP / DOWN controls where they are reversed in HA (the same you had), will try `invert_cover` tomorrow, but why this shouldn't be set in device definition? 
4) backlight is not working, don't know the reason, there was a yellow led on, while not being paired, but now I see nothing, is it the same for you?

**Comment by slashbv** [comment]
for me UP / Down works fine just that when it's opened it shows closed (0%) and when it's closed it's opened (100%). Considering Koenekk's comment on the other thread I think this should be changed too. If you use invert_cover should be fine. Maybe @1subu will fix it and he will use invert_cover for Openhab. I don't have backlight either, there is only yellow light when going Up or Down and seems the intensity can be changed with LOW MEDIUM HIGH. I think those Loratap switches don't have any backlight. I had also wifi ones from Loratap and they were the same without backlight. The one from Zemismart has backlight. Probably the one @vide has it's an improved version of the old Zemismart (the one I have) and this Loratap (it's a Loratap with Zemismart lights :)) )

**Comment by 1subu** [comment]
Regarding  cover position. 
This device reports position= 0 as open.
In devices.js you can read "coverInverted: Set to true for cover controls that report position=100 as open"
For cover position I use the existing converter cover_position_tilt. 
There are currently 15 different devices listed using this converter. Only 4 have the option coverInverted
It could not be so wrong :)

Has anyone tested the motor_reversal funktion? In the Tuya app this inverts the position, but I dont know if up/down is also inverted.

**Comment by slashbv** [comment]
it seems I have this error in Z2M log, any idea ? 

error 2020-11-09 17:54:38: Failed to setup reporting for '0xbc33acfffe47afe2' - Error: Bind 0xbc33acfffe47afe2/1 genOnOff from '0x00124b0021cc0b8a/1' failed (AREQ - ZDO - bindRsp after 10000ms)
    at Timeout._onTimeout (/app/node_modules/zigbee-herdsman/dist/utils/waitress.js:46:35)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)

**Comment by irrintxi** [comment]
Hi,
I have just received one of these LoraTap SC400 curtain switch (TUYA Zigbee model 'TS130F').
I am using Hassio on a raspberry pi 4 with a CC2531 as a zigbee hub, connected to the rasp USB.
I have installed the addon Zigbee2MQTT and the curtain switch has been discovered easily (1 device/ 6 entities).
I have added the cover entity to a card on the lovely lace.
Calibration works fine by doing as per indicated in this thread
But....this is my issue:

When I push/touch the “up” button, the curtain goes up and after the configured time the curtain stops but then the “down” button goes into disable stage and the “up” button remains enable. Shouldn’t it be vice versa?. At this point, it is not possible to move the curtain down anymore.

In the same way, if I reset everything and push/touch the “DOWN” button, the curtain goes down and after the configured time the curtain stops but then the “UP” button goes into disable stage and the “DOWN” button remains enable. Not possible to go up now…..

I am quite lost to short out this issue. Any help will be welcome.

Note: Fortunately, getting into de coder, the slider works fine.

a couple of screeshots:
https://www.dropbox.com/s/ekm6qtnxhv9tbdk/zigbee%20cover%20issue.PNG?dl=0

**Comment by slashbv** [comment]
add invert_cover: true to configuration.yaml should be fine after.

**Comment by barsurfer** [comment]
@1subu motor_reversal is also inverting up/down buttons for me. If we don't want to change coverInverted I think we are done then. Thanks for your effort!

**Comment by irrintxi** [comment]
> add invert_cover: true to configuration.yaml should be fine after.

Thanks for the response-
I forgot to mention i did already try that but no succeed... It behaves in the same manner....

**Comment by luismalddonado** [comment]
I found in an other repository (domoticz-zigbee2mqtt-plugin) how to calibrate the device:

For calibration of the switch I use next steps:

Open curtains completly
            Send calibration command to the switch: 
                        mosquitto_pub -d -m 'on' -t 'zigbee2mqtt/Living_Room_Sunblind_Switch/set/calibration'

           Press on close button on the switch, wait until curtains are fully closed and press pause
           
            Send the finalization of the calibration process:
            mosquitto_pub -d -m 'off' -t 'zigbee2mqtt/Living_Room_Sunblind_Switch/set/calibration'

**Comment by Parahassio** [comment]
Hi, complete noob here, so I would appreciate the help.

I have purchased an SM-SW101, which Home Assistant recognize as TSF130F, and I am having the same problem as shown three posts up: It reaches a moment where it seem to find the top or the bottom and the arrow deactivates.

It seem that if we add "invert_cover: true" in device.yaml solves the problem. 

But, I am completely unable to do that.
I have added in config.yaml the following
devices: devices.yaml

BUT it gives me the following error
Component error: devices - Integration 'devices' not found.

I also have created a "devices.yaml", stored in the same folder as config.yaml with the following
'0x842e14fffe9a562f':
    invert_cover: true

I have tried to paste all of that also in config.yaml, like this, with the same result
devices:
  '0x842e14fffe9a562f':
      invert_cover: true

What am I doing wrong? What can I do to solve this issue?

On the other hand, is there any way to deactivate the software ends?

Thanks for your help!!

**Comment by woodchuck84** [comment]
> Open curtains completly
> Send calibration command to the switch:
> mosquitto_pub -d -m 'on' -t 'zigbee2mqtt/Living_Room_Sunblind_Switch/set/calibration'
> Press on close button on the switch, wait until curtains are fully closed and press pause
> Send the finalization of the calibration process:
> mosquitto_pub -d -m 'off' -t 'zigbee2mqtt/Living_Room_Sunblind_Switch/set/calibration'

Tried that on zigbee2mqtt for Lonsonho Curtain Switch but in the zigbee2mqtt log I get this: No converter available for 'Calibration' (ON).
What did I mess up?

**Comment by irrintxi** [comment]
> Hi, complete noob here, so I would appreciate the help.
> 
> I have purchased an SM-SW101, which Home Assistant recognize as TSF130F, and I am having the same problem as shown three posts up: It reaches a moment where it seem to find the top or the bottom and the arrow deactivates.
> 
> It seem that if we add "invert_cover: true" in device.yaml solves the problem.
> 
> But, I am completely unable to do that.
> I have added in config.yaml the following
> devices: devices.yaml
> 
> BUT it gives me the following error
> Component error: devices - Integration 'devices' not found.
> 
> I also have created a "devices.yaml", stored in the same folder as config.yaml with the following
> '0x842e14fffe9a562f':
> invert_cover: true
> 
> I have tried to paste all of that also in config.yaml, like this, with the same result
> devices:
> '0x842e14fffe9a562f':
> invert_cover: true
> 
> What am I doing wrong? What can I do to solve this issue?
> 
> On the other hand, is there any way to deactivate the software ends?
> 
> Thanks for your help!!
> 
> Edit: The second question is because I am testing it with no motor connected, but for sure it is not on for enough time for the blind to reach the top.


Install samba shared addon, to be able to see your hassio foldes from your compuner on your local net.

![hassio net](https://user-images.githubusercontent.com/17954017/101094377-9aabb480-35bc-11eb-924f-291e5f8d82ea.png)

Then on \\HASSIO\share\zigbee2mqtt\ you will see a file named "devices.yaml". Edit it and add "invert_cover: true" on the cover device.

**Comment by Parahassio** [comment]
Thanks!

_tought that I was mistaken, because all of my devices look the same like this:

'0x842e14fffe9a562f':
  friendly_name: '0x842e14fffe9a562f'
  retain: false

So I will add there the 
  invert_cover: true

Best regards,

**Comment by irrintxi** [comment]
make sure spaces are ok. spaces are importante!
It should be like:

'0x842e14fffe9a562f':
   friendly_name: '0x842e14fffe9a562f'
   invert_cover: true

Not sure what "retain" is but I do not have it.

**Comment by Parahassio** [comment]
No, nothing. 

I have removed the retain: false but it does not work. When I move the slider up, the moving says "Down". It does not matter what I do. Maybe my device is a bit different and does not work correctly.

Is there any way to override or modify the time limit of the switch? Because at this moment is unusable, it only moves during 10 seconds and then it does not work anymore, which means that I cannot even use it as a regular switch.

**Comment by irrintxi** [comment]
Do you restart after making the changes?

**Comment by Parahassio** [comment]
Yes, I restarted the full server

**Comment by slashbv** [comment]
> it seems I have this error in Z2M log, any idea ?
> 
> error 2020-11-09 17:54:38: Failed to setup reporting for '0xbc33acfffe47afe2' - Error: Bind 0xbc33acfffe47afe2/1 genOnOff from '0x00124b0021cc0b8a/1' failed (AREQ - ZDO - bindRsp after 10000ms)
> at Timeout._onTimeout (/app/node_modules/zigbee-herdsman/dist/utils/waitress.js:46:35)
> at listOnTimeout (internal/timers.js:554:17)
> at processTimers (internal/timers.js:497:7)

Nobodyelse has this error ? Maybe I should open a new thread.

**Comment by slashbv** [comment]
I switched the log to debug and I have seen this message which seems to be related to reporting part 
debug 2020-12-09 13:40:54: No converter available for 'TS130F' with cluster 'genBasic' and type 'attributeReport' and data '{"65506":27,"appVersion":68}'

**Comment by slashbv** [comment]
@Koenkk could you take a look at this ? Can reporting be fixed or this device doesn't support reporting?

**Comment by Koenkk** [comment]
@slashbv erros like https://github.com/Koenkk/zigbee2mqtt/issues/4257#issuecomment-739848410 indicate it doesn't (so cannot be fixed from zigbee2mqtt)

**Comment by slashbv** [comment]
thanks !

**Comment by scordal** [comment]
> make sure spaces are ok. spaces are importante!
> It should be like:
> 
> ![devices](https://user-images.githubusercontent.com/17954017/101146510-5dc2da80-361b-11eb-9546-a55ef87c8b60.png)
> 
> Not sure what "retain" is but I do not have it.
> You also can change the "friendly name" for something easier to identify

- When the shutter is completelly closed, then the up button should enabled and the down button disabled. The actual result is the other way around (the down button is enabled, and the up button is disabled)
- When the shutter is completelly open, then down button should be enabled, and the up disabled. The actual result is the other way around (the up button is enabled, and the down button is disabled.

I tried by adding invert_cover: true in the devices.yaml file, in the specific cover device, and I do not get any different result.

is this actually an issue on zigbe2mqtt or home assistant? Can someone advise?

Thanks

**Comment by turbaniarz12** [comment]
Hello

I have same issue as above...I've tried everything. Help please??!!

thanks

**Comment by Parahassio** [comment]
@scordal 
To my disgrace I have purchased another switch, but this time the Lotarap SC400W-EU.

I tought that they were the same, but I am having your problem. i see the following behaviour:
1. With motor_reversal OFF
    - Position 100: Fully closed. From there:
              I can send the "open" payload and it will "moving"--> "up".
              Arrow up is greyed out, arrow down does nothing, I cannot use HA.
2. With motor_reversal ON
     - Position 0: Fully closed. From there:
             I can send the "close" payload and it will "moving" --> "down". (Towards 100).
             Arrow up is greyed out, arrow down does nothing. I cannot use HA.

How can I have the same configuration as in SM-SW101?

Thanks!

**Comment by 1subu** [comment]
The SC400W-EU is a wifi switch. The SC400ZB-EU is the zigbee switch. I hope you got the zigbee switch.

I have many of these switches in use and they work perfectly. I don't use HA, but for HA motor_reversal OFF and invert_cover: true should work.

**Comment by Parahassio** [comment]
> 
> 
> The SC400W-EU is a wifi switch. The SC400ZB-EU is the zigbee switch. I hope you got the zigbee switch.
> 
> I have many of these switches in use and they work perfectly. I don't use HA, but for HA motor_reversal OFF and invert_cover: true should work.

Hi, thanks for your comment.
You are right! SC400W-EU should be wifi and SC400ZB-EU zigbee. However, for some reason I have received a zigbee one that has the back with the wifi reference. (Zigbee is working).

However, I cannot see the way to make it work. It works as described with motor_reversal on or off: open arrow always makes the  percentage go down.

Would it be maybe that the inevrt_cover :true is not working? Is there any parameter in the configuration of zigbee2mqtt that I should modify? I just write it in devices.yaml under the correct device and with the spaces. I do not see any effect with it on or off.

Thanks for your help!

**Comment by balu79** [comment]
I'm wondering reg. the calibration, would there be any way to export / backup the calibration so that they can be restored after a power outage? I guess there isn't and won't be any such possibility. Could anybody confirm or otherwise contradict? Thanks

**Comment by yonaesp** [comment]
> I'm wondering reg. the calibration, would there be any way to export / backup the calibration so that they can be restored after a power outage? I guess there isn't and won't be any such possibility. Could anybody confirm or otherwise contradict? Thanks

I can think of a way to do it with home assistant.
Every time home assistant starts, the shutter will be placed in the configuration position and then the mqtt command will be sent to finish the configuration

**Comment by SvenVerheesen** [comment]
I have the SC400ZB-EU and added the invert_cover: true in the devices.yaml file but it does not work.
I am a beginner and maybe i am doing something wrong but it looks like Scoordal's picture.
Could anyone give me any advice what i have to change or check first?

**Comment by yonaesp** [comment]
> Me pregunto reg. la calibración, ¿habría alguna forma de exportar / hacer una copia de seguridad de la calibración para que se puedan restaurar después de un corte de energía? Supongo que no existe ni habrá tal posibilidad. ¿Alguien podría confirmar o contradecir de alguna manera? Gracias

After testing, the position settings are not lost after a power outage, so you don't have to worry.

**Comment by balu79** [comment]
Just tested it again and you are right, it doesn't seem to loose it. Not sure what happened before, but I had one case earlier where my calibration got lost and I thought it came from a temporary power outage.

Anyway, thanks for also checking from your end!

**Comment by yonaesp** [comment]
> Just tested it again and you are right, it doesn't seem to loose it. Not sure what happened before, but I had one case earlier where my calibration got lost and I thought it came from a temporary power outage.
> 
> Anyway, thanks for also checking from your end!

I read in a changelog that the data saving problem had been fixed. So I guess that was it.
I'm glad it works for you too :)

**Comment by zagi988** [comment]
Anyone here that could help me with getting a LoraTap double curtain switch to work with z2m?
I just don't understand where exactly to start with the process of adding a new device.

**Comment by dzianiwonsacze** [comment]
> Anyone here that could help me with getting a LoraTap double curtain switch to work with z2m? I just don't understand where exactly to start with the process of adding a new device.

Assuming you have Z2M installed and running, please press 'allow devices to join' button in the Z2M panel, and then either first connect the switch to mains (it will enter pairing mode automatically) or press the pairing button at the top until you see the LED blinking and Z2M will detect the device.

Once this is done, please click on the device and click 'exposes' and calibrate the device there. For some reason only there the device will actually work when calibrating, the steering panel in HA will not work until it's set up there.

I hope I helped.

**Comment by andrazek** [comment]
Any news in this regard? I also have a few double curtain switches. Hopefully, it will be possible to use it with HA someday.

**Comment by vindolin** [comment]
I also have a problem with the Loratap SC400ZB-EU.
Calibration seems to work as do the UP/DOWN commands.
When I'm controlling it manually with the physical touch buttons and stop it with the pause button, the percentage get's set to the correct value but when I use the slider in Z2M it either opens 100% or closes but nothing in between.

**Comment by joelsteinbiss** [comment]
> I also have a problem with the Loratap SC400ZB-EU.
> Calibration seems to work as do the UP/DOWN commands.
> When I'm controlling it manually with the physical touch buttons and stop it with the pause button, the percentage get's set to the correct value but when I use the slider in Z2M it either opens 100% or closes but nothing in between.

Got the same issue. Did you find a solution?

**Comment by 0rsa** [comment]
Same than @vindolin @joelsteinbiss
